File: backend/Models/Quiz.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace a15.Models;

public enum Statut
{
     EN_COURS = 0,
     FINI = 1,
     PAS_COMMENCE = 2,
     CLOTURE = 3,
     PUBLIE = 4,
     PAS_PUBLIE = 5,
}


 public class Quiz {
    [Key]
    public int Id { get; set; }
    public string Name { get; set; } = null!;
    public string? Description { get; set; }
    public bool IsPublished { get; set; }
    public bool? IsClosed { get; set; }
    public bool IsTest { get; set; }
    public DateTimeOffset? Start { get; set; }
    public DateTimeOffset? Finish { get; set; }
    [NotMapped]
    public Statut Status {get; set;}
    [NotMapped]
    public int Note {get; set;}
     [NotMapped]
    public Statut TeacherStatus {get; set;}

    [ForeignKey("DatabaseId")]
    public int DatabaseId { get; set; }
    public virtual Database Database { get; set; } = null!;

    [ForeignKey("TeacherId")]
    public int TeacherId { get; set; }
    public Teacher Teacher { get; set; } = null!;
    public virtual ICollection<Question> Questions {get; set;} = new List<Question>();
    public virtual ICollection<Answer> Answers {get; set;} = new HashSet<Answer>();
    public virtual ICollection<Attempt> Attempts {get; set;} = new HashSet<Attempt>();
    
    [NotMapped]
    public bool HasAttempt { get; set; }

    public List<string> ValidateQuizDates(Quiz quiz)
     {
     var currentDate = DateTimeOffset.UtcNow.Date; // UTC time for comparison
     var errors = new List<string>();

     if (quiz.Start.HasValue && quiz.Start.Value.UtcDateTime.Date < currentDate)
     {
          errors.Add("Start date cannot be in the past.");
     }
     if (quiz.Finish.HasValue && quiz.Finish.Value.Date < currentDate)
     {
          errors.Add("Finish date cannot be in the past.");
     }
     if (quiz.Start.HasValue && quiz.Finish.HasValue && quiz.Start.Value.Date > quiz.Finish.Value.Date)
     {
          errors.Add("Start date must be before or on the same day as the finish date.");
     }

     return errors;
     }





}

File: frontend/src/app/models/quiz.ts
import * as _ from 'lodash-es';
import { Type } from "class-transformer";
import { Question } from './question';
import { Database } from './database';


export enum Status {
    EN_COURS = 0,
    FINI = 1,
    PAS_COMMENCE = 2,
    CLOTURE=3,
    PUBLIE = 4,
    PAS_PUBLIE = 5,
}

export class Quiz {
    id?:number;
    name?: string;
    description?: string;
    isPublished?: boolean;
    isClosed?: boolean = false;
    isTest?: boolean;
    @Type(() => Date)
    start?: Date;
    @Type(() => Date)
    finish?: Date;
    database!: Database;
    questions!: Question[];
    static database: string;
    hasAttempt: boolean = false;
    note?: number;
    status?: Status | string;
    teacherStatus?: Status;

    constructor(data?: any) {
        this.name = data?.name;
        this.description = data?.description;
        this.isPublished = data?.isPublished;
        this.isTest = data?.isTest;
        this.start = data?.start;
        this.finish = data?.finish;
        this.database = data?.database;
    }

    public get getStatut(): string {
        switch (this.status) {
            case Status.PAS_COMMENCE:
                return "PAS_COMMENCE";
            case Status.EN_COURS:
                return "EN_COURS";
            case Status.CLOTURE:
                return "CLOTURE";
            case Status.FINI:
                return "FINI";
            default:
                return "PAS_COMMENCE";
        }
    }

    public get teacherStatut(): string {
        switch (this.teacherStatus) {
            case Status.PUBLIE:
                return "PUBLIE";
            case Status.PAS_PUBLIE:
                return "PAS_PUBLIE";
            case Status.CLOTURE:
                return "CLOTURE";
            default:
                return "PAS_PUBLIE";
        }
    }

    public get getNote(): string {
        //check if quiz is closed
        if (this.status === 'CLOTURE' || this.status === 'PAS_COMMENCE' || this.status ==='EN_COURS') {
            return "N/A";
        }
    
        //check if note is available
        if (this.note !== null && this.note !== undefined) {
            return this.note + "/10";
        }
        return "N/A";
    }
        
    get display(): string {
        return `${this.name} ${this.database} ${this.status}`;
    }

}

export class NewQuizDTO
{
    Id?: number | undefined;
    Name!: string;
    Description? : string;
    IsPublished!: boolean;
    IsTest!: boolean;
    Start?: Date;
    Finish?: Date;
    DatabaseId!: number;
    TeacherId!:number;
}

export interface NewQuestion {
    order: number;
    body: string;
    solutions: NewSolution[];
}

export interface NewSolution {
    order: number;
    sql: string;
}


export { Database };